racket/collects/racket/contract/private/provide.rkt

```diff
@@ -317,21 +326,19 @@
                     (quasisyntax/loc stx
                       (begin #,(internal-function-to-be-figured-out #'ctrct
                                                                     id
-                                                                    (or reflect-id id)
-                                                                    (or user-rename-id 
-                                                                        id)
+                                                                    ex-id
+                                                                    ext-name
                                                                     id-rename
                                                                     (stx->srcloc-expr srcloc-id)
                                                                     'provide/contract
                                                                     pos-module-source
                                                                     #f)
-                             #,@(let ([upe (current-unprotected-submodule-name)])
-                                  (if upe
-                                      (list #`(module+ #,upe
-                                                (provide (rename-out [#,id external-name]))))
-                                      (list)))
-                             #,@(if provide?
-                                    (list #`(provide (rename-out [#,id-rename external-name])))
+                             #,@(if upe-id
+                                    (append
+                                     (make-unprotected-submodule-code
+                                      (lambda ()
+                                        #`(provide (rename-out [#,upe-id #,ext-name]))))
+                                     (list #`(provide (rename-out [#,id-rename #,ext-name]))))
```

Comment:
I can't recall exactly what was going on here before, but why isn't there something conditional on `upe-id` here anymore? (That would be the naive change, anyway.)

Reply:
The original logic was totally unconditional: as long as the unprotected submodule exists, the `provide` *in the unprotected submodule* will be generated, even when `provide?` is false.  I’m not sure what the intent was—it’s difficult to tell because the logic was very entangled.  To simplify the matter, I just pun `upe-id` to serve double-duty as both an indicator of whether the `provide`s *in both modules* will be generated, as well as the uncontracted identifier to be exported in the unprotected submodule.  This is why both `provide`s are under the consequent.  On the other hand, `make-unprotected-submodule-code` already accounts for the case when the unprotected submodule is missing (which is probably what you really have in mind).

Reply:
Okay, I think that makes sense. Thanks!